internal/common/tree.go

```diff
@@ -25,6 +26,173 @@ type DNSTreeNodeData struct {
+	//propagate labels regardless of stop labels
+	propagateLabel(node, propLabel, value)
+
+	//propagate stop labels
+	resolveStops(node, propLabel, value, stopLabel)
+}
+
+func resolveStops(node *DNSTreeNode, label, value, stopLabel string) {
+	if isRoot(node) && allChildrenHaveLabel(node, label, value) {
+		RemoveLabelFromTree(node, label)
+		//entire tree cleaned so no need for any further checks
+		return
+	}
+
+	//remove label from stop labelled children
+	for _, c := range node.Children {
+		d := findDataSetForChild(node, c.Name)
+		//has label and stop label = remove label from this dataset and all children
+		if d != nil && d.Labels[stopLabel] != "" && d.Labels[label] != "" {
+			delete(d.Labels, label)
+			RemoveLabelFromTree(c, label)
+			RemoveLabelFromParents(node, label)
+		} else {
+			resolveStops(c, label, value, stopLabel)
+		}
+
+	}
+
+}
+
+func propagateLabel(node *DNSTreeNode, label, value string) {
+	for _, c := range node.Children {
+		d := findDataSetForChild(node, c.Name)
+		if d != nil && d.Labels[label] != "" {
+			//this child is labelled, indiscriminately label entire tree under this child
+			AddLabelToTree(c, label, value)
+		} else {
+			// this child is not labelled, continue descending to propagate label
+			propagateLabel(c, label, value)
+		}
+	}
+
+	// if all children are labelled, label this branch in parent node
+	if len(node.Children) > 0 && allChildrenHaveLabel(node, label, value) && node.Parent != nil {
+		AddLabelToBranch(node.Parent, node.Name, label, value)
+	}
+}
+
+func isRoot(node *DNSTreeNode) bool {
+	return node.Parent == nil
+}
+
+func allChildrenHaveLabel(node *DNSTreeNode, label, value string) bool {
+	for _, c := range node.Children {
+		if !HasLabelForBranch(node, c.Name, label, value) {
+			return false
+		}
+	}
+	return true
+}
+
+func findDataSetForChild(node *DNSTreeNode, name string) *DNSTreeNodeData {
+	for _, d := range node.DataSets {
+		if slices.Contains(d.Targets, name) {
+			return &d
+		}
+	}
+	return nil
+}
+
+func AddLabelToBranch(node *DNSTreeNode, branch, label, value string) {
+	d := findDataSetForChild(node, branch)
+	if d == nil {
+		node.DataSets = append(node.DataSets, DNSTreeNodeData{
+			Labels: endpoint.Labels{
+				label: value,
+			},
+			Targets: []string{
+				branch,
+			},
+		})
+	} else {
+		if len(d.Targets) == 1 {
+			d.Labels[label] = value
+		} else {
+			//remove target from shared dataset and recreate uniquely
+			for i, t := range d.Targets {
+				if t == branch {
+					d.Targets = append(d.Targets[:i], d.Targets[i+1:]...)
+					newDS := DNSTreeNodeData{
+						Labels:  d.Labels.DeepCopy(),
+						Targets: []string{branch},
+					}
+					newDS.Labels[label] = value
+					node.DataSets = append(node.DataSets, newDS)
+				}
+			}
+		}
+	}
+}
+
+func AddChild(parent *DNSTreeNode, child *DNSTreeNode) {
+	parent.Children = append(parent.Children, child)
+	child.Parent = parent
+}
+
+func RemoveLabelFromParents(node *DNSTreeNode, label string) {
```

Comment:
Wondering if there is a bettter name as it is not obvious that root is not a parent? Just reading this name I would expect it removes it from all parents

`RemoveLabelFromNonRootParents`?

Reply:
well, it's only returning early if the node is a root, as a root doesn't have parents (Which is all that `IsRoot` tests for).

Or in other words, if the node you've passed in to `RemoveLabelFromParents`  has no parents then there is no work to do.

Reply:
ah, it shouldn't be public, I'll update that